server/src/ast/mod.rs

```diff
@@ -34,15 +40,302 @@ pub enum SyntaxTree {
+        match node.tree {
+            SyntaxTree::Variable(value) => Values::Variable(value),
+            SyntaxTree::Number(value) => Values::Number(value),
+            SyntaxTree::Sum(box left, box right) => Values::from(left).add(right.into()),
+            SyntaxTree::Product(box left, box right) => Values::from(left).mul(right.into()),
+            SyntaxTree::Exponent(box left, box right) => Values::from(left).exp(right.into()),
+            SyntaxTree::Subtraction(box left, box right) => Values::from(left)
+                .add(Values::from(right).mul(Values::Number(BigDecimal::from(-1)))),
+                
+            SyntaxTree::Division(box left, box right) => Values::from(left)
+                .mul(Values::from(right).exp(Values::Number(BigDecimal::from(-1)))),
+            SyntaxTree::Negation(box value) => {
+                Values::from(value).mul(Values::Number(BigDecimal::from(-1)))
+            }
+        }
+    }
+}
+
+impl Display for Values {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Values::Variable(value) => write!(f, "{}", value),
+            Values::Number(value) => write!(f, "{}", value),
+            Values::Sum(constant, vmap) => {
+                write!(f, "( {constant} ").unwrap();
+                for (k, v) in vmap {
+                    write!(f, "+ {v} * {k} ").unwrap();
+                }
+                write!(f, ")")
+            }
+            Values::Product(constant, vmap) => {
+                write!(f, "( {constant} ").unwrap();
+                for (k, v) in vmap {
+                    write!(f, "* {k} ^ {v} ").unwrap();
+                }
+                write!(f, ")")
+            }
+            Values::Exponent(box (left, right)) => write!(f, "({} ^ {})", left, right),
+        }
+    }
+}
+//const one : BigDecimal = BigDecimal::from(1);
+
+//const two : BigDecimal = BigDecimal::from(2);
+
+fn exponantiate(mut x: BigDecimal, mut y: BigDecimal) -> BigDecimal {
+    if (&y).is_integer() {
+        if (&y).rem(BigDecimal::from(2)) == BigDecimal::zero() {
+            let n = y.half();
+
+            while (&y) > &n {
+                x = x.square();
+                y -= BigDecimal::one();
+            }
+            x
+        } else {
+            let n = (y.clone() - BigDecimal::one()).half();
+            let m = x.clone();
+            while (&y) > &n {
+                x = x.square();
+                y -= BigDecimal::one();
+            }
+            x * m
+        }
+    } else {
+        panic!("not implemented yet")
+    }
+}
+
+impl Values {
+    // fn entry_adder(
+    //     mut map: BTreeMap<Values, BigDecimal>,
+    //     pair: (&Values, &BigDecimal),
+    // ) -> BTreeMap<Values, BigDecimal> {
+    //     let (key, value) = pair;
+    //     map.entry(key.to_owned())
+    //         .and_modify(|x| *x += value)
+    //         .or_insert(value.to_owned());
+    //     map
+    // }
+
+    fn entry_adder<F>(
```

Comment:
this is unnecessarily general. i had a brain fart. commented out one should work